### PR TITLE
[SPARK-27108][SQL] Add parsed SQL plans for create, CTAS.

### PR DESCRIPTION
**Pull request by rdblue**
## What changes were proposed in this pull request?

This moves parsing `CREATE TABLE ... USING` statements into catalyst. Catalyst produces logical plans with the parsed information and those plans are converted to v1 `DataSource` plans in `DataSourceAnalysis`.

This prepares for adding v2 create plans that should receive the information parsed from SQL without being translated to v1 plans first.

This also makes it possible to parse in catalyst instead of breaking the parser across the abstract `AstBuilder` in catalyst and `SparkSqlParser` in core.

For more information, see the [mailing list thread](https://lists.apache.org/thread.html/54f4e1929ceb9a2b0cac7cb058000feb8de5d6c667b2e0950804c613@%3Cdev.spark.apache.org%3E).

## How was this patch tested?

This uses existing tests to catch regressions. This introduces no behavior changes.

### CODE REVIEW

**Comment by rdblue** [comment]
@cloud-fan, this is needed to add the v2 create and CTAS plans. We can get a start while waiting for the catalog identifiers to be committed.

**Comment by rdblue** [comment]
cc @jzhuge, @mccheah, @gatorsmile

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala]: set this to final at the very least

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala]: not sure if this is useful hierarchy, but if yes we should document more clearly this should not survive analysis.

Comment (rxin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala]: might be useful to add a special check for this, rather than relying on resolved only.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/CreateTable.scala
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/CreateTable.scala]: make it clear here that create table has no data operation. purely metadata.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: just to double check, when did we deprecate `CreateTempViewUsing`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: I didn't review this file carefully, assuming it's just copy-paste code from `SparkSqlAstBuilder`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala]: is it really necessary to have this parent class just to set the `resolved` bit? I think we can just put `override lazy val resolved = false` in the new `CreateTable` and `CreateTableAsSelect` classes, with classdoc saying that these 2 classes will be replaced by what concrete plans during analysis.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Yes, this is moving what is needed from `SparkSqlAstBuilder`. The only real changes are in the `visitCreateTable` method.

These rules should have already been in the abstract class. I'm not sure why they were in `SparkSqlAstBuilder` other than that was the easiest place to put them when they were added.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala]: The value of this class is that it identifies the set of logical plans that correspond directly to what was parsed from SQL. When someone working on a plan sees `ParsedLogicalPlan` as an ancestor in Scaladoc, it signals what is explained here: that parser produces `ParsedLogicalPlan` nodes without translating what was parsed, then those plans get translated into real plans in the analyzer.

With that information, it is easy to see what changes need to be made. If the parsed plan doesn't include an option, then the parser and parsed plan needs to be updated. If it does include an option, then the analyzer and downstream plans need to be updated.

Also keep in mind that this is the first two subclasses of `ParsedLogicalPlan`. To implement v2 along-side v1, we are going to be adding more of them. So it is valuable that we don't need to remember to set `resolved` to false in every plan.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Looks like it happened in 5effc016c893ce917d535cc1b5026d8e4c846721, 3 years ago.

We can always create a parsed plan for `CreateTempViewUsing` so that it can move to catalyst as well, but I thought that we can do it later, and only need to if this isn't going to be removed in 3.0.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala]: Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/CreateTable.scala
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/CreateTable.scala]: Done.

**Comment by rdblue** [comment]
@cloud-fan, I've responded to the review comments and implemented fixes.

Also, I moved the new resolution rules into `DataSourceResolution`. The new rules needed to be added to `extendedResolutionRules` instead of `postHocResolutionRules` because the post-hoc rules are only run once. This resolution needs to be done before the post-hoc rules defined in `DataSourceAnalysis` run.

At a minimum, these needed to be separated into different classes, but I think it is also more correct for resolution rules to run in the resolution batch so that other resolution rules can run on the plans produced by these rules. This is the same reason why we added the `ResolveOutputRelation` rule to the resolution batch.

I think that the resolution rules in `DataSourceAnalysis` should also move into the resolution batch, but I think that should be done as a separate change.

**Comment by rdblue** [comment]
Retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: ah, so we deprecated `CREATE TABLE USING`, but not `CREATE TEMP VIEW USING`

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM if tests pass

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala]: why are some test cases duplicated?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: shall we call it `DDLResolution`? It's not very related to data source.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala
Comment (rdblue) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala]: These were split into two test suites:

* A DDL suite that validates SQL conversion to the parsed logical plan
* A DDL/Resolution suite that runs the exact same tests to validate that the extra conversion step doesn't change anything

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: This resolves parsed plans to `execution.datasources` plans. This isn't just for DDL. We are starting with `CreateTable` and `CreateTableAsSelect`, but there will be more parsed plans that get converted to datasource plans in this rule. That's why I think `DataSourceResolution` is an appropriate name.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala]: shall we merge these 2 test suites? Or we need to add document to say that, when adding a new DDL command, we need to add a test in both `DDLParserSuite` and `PlanResolutionSuite` for different purpose.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala]: shall we merge these 2 test suites? Or we need to add document to say that, when adding a new DDL command, we need to add a test in both `DDLParserSuite` and `PlanResolutionSuite` for different purpose.

Comment (rdblue) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala]: No.

These suites are in separate modules. The parser suite is in catalyst, where new tests will go. The old tests that validate there are no changes when converting SQL -> parsed plan -> datasource plan must be in core because the plans don't exist in catalyst.

There is no need to add documentation because these will evolve separately. New parsed plans for DDL will be added to the DDL suite. Conversions will be tested when they are added. For example, we are going to add DSv2 plans for create and CTAS. Those conversions will be tested independently.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala
Comment (mccheah) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala]: +1 - @rdblue these should only be inputs to the analyzer, not outputs. Would be helpful to write specific JavaDoc on this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/CreateTable.scala
Comment (mccheah) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/CreateTable.scala]: Just so I understand correctly, `DDLResolution` returns an instance of `org.apache.spark.sql.execution.datasources.CreateTable`, which is still a subtype of `LogicalPlan`. It's a little confusing at least to me to have two `LogicalPlan` types be named the same thing, but they represent different steps of the query planning, and are in different packages.

What if we called this variant of `CreateTable`, say, making this `ParsedCreateTable` or perhaps `CreateTableStatement`?

Also, when we start partitioning between V1 sources and V2 sources in the Analyzer, are we going to return the same type of `LogicalPlan` back - a `org.apache.spark.sql.execution.datasources.CreateTable` regardless of it's loading V1 or V2? Or should we instead be naming `org.apache.spark.sql.execution.datasources.CreateTableV1`, anticipating the Analyzer to return instances of `CreateTableV2` plans down the road?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala]: That is included above: "Parsed logical plans are not resolved because they must be converted to concrete logical plans."

Do you think that should be rephrase to be more clear?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/CreateTable.scala
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/CreateTable.scala]: I'm okay using packages to differentiate between these. The rules use `sql.CreateTable` and I think that is clear.

I'd also like to avoid renaming the v1 classes to include `V1` in their names. That's going to create a lot of conflicts when working with branches and it's hard enough to maintain PRs already given the amount of needless renaming, whitespace changes, and import changes happening.

Similarly, I want to avoid adding `V2` to class names because we eventually want to remove the old packages and support. Then we would be left with a lot of v2 class names, or we would have to refactor to rename. Using packages as qualifiers is well supported in Scala, so I like to use that for this.

Comment (mccheah) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/CreateTable.scala]: In the future then when we do split between `v1` and `v2`, perhaps we can clarify by qualifying just the package? Though I guess that goes against Spark convention, but something like this:

```
import org.apache.spark...v1
import org.apache.spark...v2

...

return v1.CreateTable(...)
...
return v2.CreateTable(...)
```
Otherwise we could alias the imports:

```
import org.apache.spark....v1.{CreateTable => CreateTableV1}
import org.apache.spark....v2.{CreateTable => CreateTableV2}
```

Whatever the case then I think disambiguating by package is fine.

Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/CreateTable.scala]: Sounds good to me! I'll leave this as is for now.

We can also bring this up in the sync to see if there are other opinions.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala
Comment (mccheah) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala]: Defer to @rxin but I'm ok with merging with the current docs. We can rephrase in a follow-up if our contributors have trouble with this wording.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/CreateTable.scala
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/CreateTable.scala]: > I'm like using packages to differentiate between the plans.

This reminds me that we add `Exec` postfix to all physical plans, because we think it's not ideal to differentiate plans only be package name. When you search a class by name with an IDE, it will take you more time to pick a class with different package names.

`CreateTableStatemet` sounds good to me.

**Comment by rdblue** [comment]
@mccheah, @cloud-fan, I've rebased on master, fixed a minor conflict, and added the `Statement` suffix to the new plans. I think this should be good to go once tests pass.

**Review by mccheah — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by rdblue** [comment]
Thanks @cloud-fan! And thanks to all the reviewers also!

**Pull request closed without merging**
